### PR TITLE
Faster module name lookup

### PR DESCRIPTION
**Pull request by slavapestov**
Builds on https://github.com/apple/swift/pull/26864 by replacing lookupInModule() with a much simpler implementation that walks a cached traversal of the import graph.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
https://github.com/apple/swift-lldb/pull/1940
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
apple/swift-lldb#1940
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
apple/swift-lldb#1940
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: lib/AST/ModuleNameLookup.cpp [hunk 1]
```diff
@@ -254,194 +114,114 @@ class LookupVisibleDecls : public ModuleNameLookup<LookupVisibleDecls> {
...
+
+  bool canReturnEarly = (initialCount != decls.size() &&
+                         getDerived()->canReturnEarly());
+  if (canReturnEarly &&
+      resolutionKind == ResolutionKind::Overloadable) {
+    // If we only found top-level functions, keep looking, since we may
+    // find additional overloads.
+    if (std::find_if(decls.begin() + initialCount, decls.end(),
+                     [](ValueDecl *VD) { return !isa<FuncDecl>(VD); })
+        == decls.end())
```
Comment (jrose-apple) [inline lib/AST/ModuleNameLookup.cpp]: Nitpick: `std::none_of`

File: lib/AST/ModuleNameLookup.cpp [hunk 2]
```diff
@@ -254,194 +114,114 @@ class LookupVisibleDecls : public ModuleNameLookup<LookupVisibleDecls> {
...
+      }
     }
 
-    auto &resultSet = next.first.empty() ? unscopedValues : scopedValues;
-    lookupInModule(resultSet, next.second, combinedAccessPath,
-                   moduleScopeContext);
-  }
+    for (auto import : imports.getTopLevelImports()) {
+      // A module appears in its own top-level import list; since we checked
+      // it already, skip it.
```
Comment (jrose-apple) [inline lib/AST/ModuleNameLookup.cpp]: It shouldn't.

File: lib/AST/ModuleNameLookup.cpp [hunk 3]
```diff
@@ -254,194 +114,114 @@ class LookupVisibleDecls : public ModuleNameLookup<LookupVisibleDecls> {
...
+    }
   }
 
+  // Nothing more to do if we don't have ambiguous results.
+  if (decls.size() - initialCount <= 1)
+    return;
+
   // Remove duplicated declarations, which can happen when the same module is
-  // imported indirectly through two intermediate modules.
+  // imported with multiple access paths.
```
Comment (jrose-apple) [inline lib/AST/ModuleNameLookup.cpp]: I don't think that's correct. Then you wouldn't get duplicates.

File: lib/Sema/TypeCheckAttr.cpp [hunk 4]
```diff
@@ -1612,15 +1613,13 @@ void AttributeChecker::checkApplicationMainAttribute(DeclAttribute *attr,
...
-                                           SourceLoc(),
-                                           lookupOptions);
-    if (lookup.size() == 1)
-      ApplicationDelegateProto = dyn_cast<ProtocolDecl>(
-        lookup[0].getValueDecl());
+    SmallVector<ValueDecl *, 1> decls;
+    namelookup::lookupInModule(KitModule, Id_ApplicationDelegate,
+                               decls, NLKind::QualifiedLookup,
+                               namelookup::ResolutionKind::TypesOnly,
+                               KitModule);
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckAttr.cpp]: This can use `nullptr` as the context, meaning "only public stuff".

File: lib/AST/ModuleNameLookup.cpp (hunk 3 above)
Comment (slavapestov) [reply lib/AST/ModuleNameLookup.cpp]: What if you have "import Foo; import Foo.Bar"?

File: lib/AST/ModuleNameLookup.cpp (hunk 2 above)
Comment (slavapestov) [reply lib/AST/ModuleNameLookup.cpp]: To be clear that comment refers to an artifact of how ImportSet is built. It's not talking about what getImportedModules() returns.

When building an ImportSet we compute the transitive closure of a set of ImportedModules. If you ask for the ImportSet of a source file, the generating set contains the module itself and the source file's imports. If you ask for the ImportSet of a module, the generating set is just the module itself.

File: lib/AST/ModuleNameLookup.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/AST/ModuleNameLookup.cpp]: Oops, I missed that one. Thanks!

File: lib/AST/ModuleNameLookup.cpp (hunk 3 above)
Comment (jrose-apple) [reply lib/AST/ModuleNameLookup.cpp]: Oh right.

File: lib/AST/ModuleNameLookup.cpp (hunk 2 above)
Comment (jrose-apple) [reply lib/AST/ModuleNameLookup.cpp]: Ahh, okay. I guess that makes sense.

**Pull request merged**